lib/hd/keyorigin.js

```diff
@@ -0,0 +1,173 @@
+/*!
+ * keyorigin.js - hd key path object for bcoin
+ * Copyright (c) 2023, the bcoin developers (MIT License).
+ * https://github.com/bcoin-org/bcoin
+ */
+
+'use strict';
+
+const assert = require('assert');
+const common = require('./common');
+const bio = require('bufio');
+const {inspectSymbol} = require('../utils');
+/**
+ * Helper class to represent hd key path for arbitrary wallets.
+ * @property {Number} fingerPrint - master key fingerprint (uint32)
+ * @property {Array} path - bip32 derivation path in uint32 array
+ */
+class KeyOriginInfo {
+  constructor(options) {
+    this.fingerPrint = -1;
+    this.path = [];
+    if (options) {
+      this.fromOptions(options);
+    }
+  }
+
+  fromOptions(options) {
+    assert(options, 'requires options');
+
+    if (options.fingerPrint !== undefined) {
+      assert(
+        options.fingerPrint >>> 0 === options.fingerPrint &&
+          options.fingerPrint <= 0xffffffff,
+        'fingerPrint must be uint32'
+      );
+      this.fingerPrint = options.fingerPrint;
+    }
+
+    if (options.path !== undefined) {
+      if (Array.isArray(options.path)) {
+        assert(
+          options.path.every(p => p >>> 0 === p && p <= 0xffffffff),
+          'all path index must be uint32'
+        );
+        this.path = options.path;
+      } else {
+        this.path = common.parsePath(options.path, true);
+      }
+    }
+    return this;
+  }
+
+  static fromOptions(options) {
+    return new this().fromOptions(options);
+  }
+
+  equals(keyInfo) {
+    assert(KeyOriginInfo.isKeyOriginInfo(keyInfo));
+    if (this.fingerPrint !== keyInfo.fingerPrint) {
+        return false;
+    }
+    assert(this.path.length === keyInfo.path.length);
+    for (let i = 0; i < this.path.length; i++) {
+      if (this.path[i] !== keyInfo.path[i]) {
+            return false;
+      }
+    }
+    return true;
+  }
+
+  inspect() {
+    return this.format();
+  }
+
+  format() {
+    let path = 'm';
```

Comment:
Is `m` always hardened? i.e. `m'`?

Reply:
`m` just specifies the key from which derivation starts. It doesn't need to be `m'` in case the key is hardened.
If we have a key say `xpub....`, and the derivation path is `m/0`, it means derive the unhardened child at index `0`, if the path is `m/0'`, it means derive the hardened child at index `0` (in this case if `m` is not hardened, the derivation will fail).